cli/cliui/prompt.go

```diff
@@ -179,3 +178,29 @@ func promptJSON(reader *bufio.Reader, line string) (string, error) {
 	}
 	return line, nil
 }
+
+// readUntil the first occurrence of delim in the input, returning a string containing the data up
+// to and including the delimiter. Unlike `bufio`, it only reads until the delimiter and no further
+// bytes. If readUntil encounters an error before finding a delimiter, it returns the data read
+// before the error and the error itself (often io.EOF). readUntil returns err != nil if and only if
+// the returned data does not end in delim.
+func readUntil(r io.Reader, delim byte) (string, error) {
+	var (
+		have []byte
+		b    = make([]byte, 1)
+	)
+	for {
+		n, err := r.Read(b)
+		if n > 0 {
```

Comment:
Minor note: since `r.Read` blocks until it receives content or encounters an error, could this be restructured perhaps as follows?

```
	for {
		_, err := r.Read(b)
		if err != nil {
			return string(have), err
		}
		have = append(have, b[0])
		if b[0] == delim {
			return string(have), nil
		}
	}
```

Reply:
from the docs for `io.Reader.Read`

> When Read encounters an error or end-of-file condition after successfully reading n > 0 bytes, it returns the number of bytes read. It may return the (non-nil) error from the same call or return the error (and n == 0) from a subsequent call. An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF.
>
> **Callers should always process the n > 0 bytes returned before considering the error err.** Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors.

(emphasis mine)